server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndexer.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.component.index;
+
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.Uninterruptibles;
+import java.util.Arrays;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.elasticsearch.action.index.IndexRequest;
+import org.sonar.api.Startable;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.server.es.BulkIndexer;
+import org.sonar.server.es.EsClient;
+
+import static org.elasticsearch.index.query.QueryBuilders.termQuery;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_PROJECT_UUID;
+import static org.sonar.server.component.index.ComponentIndexDefinition.INDEX_COMPONENTS;
+import static org.sonar.server.component.index.ComponentIndexDefinition.TYPE_COMPONENT;
+
+public class ComponentIndexer implements Startable {
+
+  private final ThreadPoolExecutor executor;
+  private final DbClient dbClient;
+  private final EsClient esClient;
+
+  public ComponentIndexer(DbClient dbClient, EsClient esClient) {
+    this.executor = new ThreadPoolExecutor(0, 1, 0L, TimeUnit.SECONDS, new LinkedBlockingQueue<>());
+    this.dbClient = dbClient;
+    this.esClient = esClient;
+  }
+
+  /**
+   * Copy all components of all projects to the elastic search index.
+   * <p>
+   * <b>Warning</b>: This should only be called on an empty index. It does not purge anything.
+   */
+  public void index() {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      dbClient.componentDao()
+        .selectProjects(dbSession)
```

Comment:
Only UUIDs of ComponentDto are needed. Moreover all the projects are loaded in memory and the Elasticsearch bulk index requests may not be fully populated when projects are too small.
I suggest to iterate over the cursor of all enabled components:
```
try (DbSession dbSession = dbClient.openSession(false)) {
      BulkIndexer bulk = new BulkIndexer(esClient, INDEX_COMPONENTS);
      // ...
      dbClient.componentDao()
        .selectAll(dbSession, context -> {
          ComponentDto dto = (ComponentDto) context.getResultObject();
          bulk.add(newIndexRequest(toDocument(dto)));
        });
      // ...
    }
```